includes/Hooks.php

```diff
@@ -2,95 +2,125 @@
-     *
-     * @param OutputPage   $output
-     * @param Article      $article
-     * @param Title        $title
-     * @param User         $user
-     * @param WebRequest   $request
-     * @param MediaWiki    $wiki
-     * @return bool        False to abort further action
-     */
-    public static function onMediaWikiPerformAction(
-        OutputPage $output,
-        Article $article,
-        Title $title,
-        User $user,
-        WebRequest $request,
-        MediaWiki $wiki
-    ) {
-        $type   = $request->getVal( 'type' );
-        $action = $request->getVal( 'action' );
-        $diffId = (int)$request->getVal( 'diff' );
-        $oldId  = (int)$request->getVal( 'oldid' );
+		if (
+			!$user->isRegistered()
+			&& (
+				$type === 'revision'
+				|| $action === 'history'
+				|| $diffId > 0
+				|| $oldId > 0
+			)
+		) {
+			$this->denyAccess( $output );
+			return false;
+		}
 
-        if (
-            !$user->isRegistered()
-            && (
-                $type === 'revision'
-                || $action === 'history'
-                || $diffId > 0
-                || $oldId  > 0
-            )
-        ) {
-            self::denyAccess( $output );
-            return false;
-        }
+		return true;
+	}
 
-        return true;
-    }
+	/**
+	 * Block Special:RecentChangesLinked and Special:WhatLinksHere for anonymous users.
+	 *
+	 * @param SpecialPage $special
+	 * @param string|null $subPage
+	 * @return bool False to abort execution
+	 */
+	public function onSpecialPageBeforeExecute( $special, $subPage ) {
+		$user = $special->getContext()->getUser();
+		if ( $user->isRegistered() ) {
+			// logged-in users: allow
+			return true;
+		}
 
-    /**
-     * Block Special:RecentChangesLinked and Special:WhatLinksHere for anonymous users.
-     *
-     * @param SpecialPage $specialPage
-     * @param string      $subPage
-     * @return bool       False to abort execution
-     */
-    public static function onSpecialPageBeforeExecute( SpecialPage $specialPage, $subPage ) {
-        $user = $specialPage->getContext()->getUser();
-        if ( $user->isRegistered() ) {
-            return true; // logged-in users: allow
-        }
+		$name = strtolower( $special->getName() );
+		if ( in_array( $name, [ 'recentchangeslinked', 'whatlinkshere' ], true ) ) {
+			$out = $special->getContext()->getOutput();
+			$this->denyAccess( $out );
+			return false;
+		}
 
-        $name = strtolower( $specialPage->getName() );
-        if ( in_array( $name, [ 'recentchangeslinked', 'whatlinkshere' ], true ) ) {
-            $out = $specialPage->getContext()->getOutput();
-            self::denyAccess( $out );
-            return false;
-        }
+		return true;
+	}
 
-        return true;
-    }
+	/**
+	 * Helper: output 403 Access Denied page using i18n messages.
+	 *
+	 * @param OutputPage $output
+	 * @return void
+	 */
+	protected function denyAccess( OutputPage $output ): void {
```

Comment:
Any specific reason this is now protected instead of private? I'm not too familiar with any PHP-specific conventions about visibility, so any further info on this would be appreciated.

Reply:
It's protected so that it can be mocked away via phpunit. phpunit mocking works by dynamically extending the class being mocked and overriding specific methods. By making the method protected, we are able to replace it in the unit tests with a dummy version that simply returns `null` instead of running the actual logic inside of the method, which breaks in unit testing because calling `wfMessage(...)->plain()` or `->text()` or the like requires the service container.

In an integration test the service container is defined and we can use the full message framework, but an integration test should also be using actual User, WebRequest, etc. objects rather than mocks.